src/Strings.h

```diff
@@ -390,97 +370,52 @@ void ParseNgramWord(std::string &input,
-
-    if (ngramRangeMin < 1 || ngramRangeMin > ngramRangeMax || ngramRangeMax > trimedString.size())
+    if (ngramRangeMin < 1 || ngramRangeMin > ngramRangeMax || ngramRangeMax > input.size())
         throw std::invalid_argument("ngramRangeMin and ngramRangeMax not valid");
 
-    Details::ParseNgramCharHelper<IteratorT>(trimedString.begin(), trimedString.end(), ngramRangeMin, ngramRangeMax, callback);
+    Details::ParseNgramCharHelper<IteratorT>(input.begin(), input.end(), ngramRangeMin, ngramRangeMax, callback);
 }
 
-template <typename IteratorT>
-void ParseNgramCharCopy(std::string const &input,
-                        size_t const ngramRangeMin,
-                        size_t const ngramRangeMax,
-                        std::function<void (IteratorT, IteratorT)> const &callback) {
-
-    //copy string to trim
-    std::string inputCopy(input);
-    ParseNgramChar(inputCopy, ngramRangeMin, ngramRangeMax, callback);
-}
 
 template <
     typename IteratorT,
     typename UnaryPredicateT
 >
-void ParseNgramCharwb(std::string &input,
+void ParseNgramCharwb(std::string const &input,
                       UnaryPredicateT const &predicate,
                       size_t const ngramRangeMin,
                       size_t const ngramRangeMax,
                       std::function<void (IteratorT, IteratorT)> const &callback) {
 
-    std::string trimedString(Details::ReplaceAndDeDuplicate<std::function<bool (char)>>(input));
-
-    //wordIterPairVector is used to store the begin and end iterator of words in input
     std::vector<std::pair<IteratorT, IteratorT>> wordIterPairVector;
-    Details::Parse<IteratorT, UnaryPredicateT>(
-        trimedString.begin(),
-        trimedString.end(),
-        predicate,
-        [&wordIterPairVector] (IteratorT iterBegin, IteratorT iterEnd) {
-            wordIterPairVector.emplace_back(iterBegin, iterEnd);
-        }
-    );
 
-    if (wordIterPairVector.size() == 0)
-        throw std::invalid_argument("wordIterPairVector.size() == 0");
+    IteratorT itr = input.begin();
+    IteratorT end = input.end();
+
+    while (itr != end) {
+        if (predicate(*itr)) {
+            IteratorT spaceItr = itr;
+            IteratorT wordBeginItr = ++itr;
+            wordIterPairVector.emplace_back(spaceItr, wordBeginItr);
+        } else {
+            ++itr;
+        }
+    }
 
     if (ngramRangeMin < 1 || ngramRangeMin > ngramRangeMax )
         throw std::invalid_argument("ngramRangeMin and ngramRangeMax not valid");
 
-    for (size_t pairIdx = 0; pairIdx < wordIterPairVector.size(); ++pairIdx) {
-        //copy
-        std::string const paddedWord = " " + std::string(wordIterPairVector[pairIdx].first, wordIterPairVector[pairIdx].second) + " ";
-
-        Details::ParseNgramCharHelper<IteratorT>(paddedWord.begin(), paddedWord.end(), ngramRangeMin, ngramRangeMax, callback);
+    for (size_t pairIdx = 0; pairIdx < wordIterPairVector.size() - 1; ++pairIdx) {
```

Comment:
This code strikes me as a little bit odd, as it using using the first and second values from different iterators. Is this intended? If so, add a comment explaining what is happening.

Reply:
comments added